# Dataquest Guided Project: Profitable App Profiles for the App Store and Google Play Markets
Lev Craig ([levcb](www.github.com/levcb))

This project identifies mobile app profiles that are profitable in the App Store and Google Play markets. This project provides data insights intended to facilitate data-driven decisions about app characteristics for a hypothetical company building free apps. Accordingly, this hypothetical company's main source of revenue is in-app ads, meaning that the revenue for a given app is driven primarily by its number of users.

Two datasets are used here: one containing data about approximately 10,000 Android apps from Google Play, and another containing data about approximately 7,000 iOS apps from the App Store.

We'll start by opening and exploring the two datasets. The below function prints rows in a readable way.

In [1]:
def explore_data(dataset, start, end, rows_and_columns=False):
    # define section of dataset to explore
    dataset_slice = dataset[start:end]
    # print each row in selected slice
    for row in dataset_slice:
        print(row)
        # insert line break to improve readability
        print('\n')
    # if rows_and_columns parameter set to True, print numbers of rows and columns
    if rows_and_columns:
        print("Number of rows: ", len(dataset))
        print("Number of columns: ", len(dataset[0]))

Next, we'll open each dataset.

In [2]:
from csv import reader

# open Apple Store dataset
ios_data = list(reader(open('AppleStore.csv')))
# separate header row from data rows
ios_header = ios_data[0]
ios_data = ios_data[1:]

# open Google Play dataset
android_data = list(reader(open('googleplaystore.csv')))
# separate header row from data rows
android_header = android_data[0]
android_data = android_data[1:]

Now that we've opened the datasets, we'll use our `explore_data` function to explore them. The below cells print the first few rows and the number of columns and rows in each dataset.

In [3]:
# print first three rows and number of rows and columns in iOS dataset
explore_data(ios_data, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows:  7197
Number of columns:  16


In [4]:
# print first three rows and number of rows and columns in Android dataset
explore_data(android_data, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows:  10841
Number of columns:  13


Now we'll print the column headings for each dataset to help us identify which columns will be most useful for our purposes. (To make this data more comprehensible, I've created tables providing the iOS and Android column names and descriptions.)

In [5]:
print(ios_header)
print(android_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


### iOS
| Column Name        | Description                                   |
|--------------------|-----------------------------------------------|
| `id`               | App ID                                        |
| `track_name`       | App name                                      |
| `size_bytes`       | Size (in bytes)                               |
| `currency`         | Currency type                                 |
| `price`            | Price amount                                  |
| `rating_count_tot` | User rating counts (for all versions)         |
| `rating_count_ver` | User rating counts (for current version)      |
| `user_rating`      | Average user rating (for all versions)        |
| `user_rating_ver`  | Average user rating (for current version)     |
| `ver`              | Latest version code                           |
| `cont_rating`      | Content rating                                |
| `prime_genre`      | Primary genre                                 |
| `sup_devices.num`  | Number of supporting devices                  |
| `ipadSc_urls.num`  | Number of screenshots shown for display       |
| `lang.num`         | Number of supported languages                 |
| `vpp_lic`          | Whether VPP device-based licensing is enabled |

### Android
| Column Name        | Description                                          |
|--------------------|------------------------------------------------------|
| `App`              | App name                                             |
| `Category`         | Category the app belongs to                          |
| `Rating`           | Overall user rating of the app                       |
| `Reviews`          | Number of user reviews for the app                   |
| `Size`             | Size of the app                                      |
| `Installs`         | Number of user downloads/installs                    |
| `Type`             | Paid or free                                         |
| `Price`            | Price of the app                                     |
| `Content Rating`   | Age group the app is targeted at                     |
| `Genres`           | Genres an app belongs to in addition to its category |


Before beginning our analysis, we'll remove any incorrect or duplicate data, any non–English language apps, and any apps that aren't free (as we're only interested in English apps that are free to download).

Other users have found that the Google Play dataset has a mistake in row 10472:

In [6]:
print(android_data[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The `Category` value (`android_data[10472][8]`) is missing from this entry, which causes a column shift.

In the next code cell, we'll delete this row to remove the problem data from our dataset.

In [7]:
del android_data[10472]

Now, if we print `android_data[10472]` again, we'll see that the problem row has been deleted:

In [8]:
print(android_data[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


The Google Play dataset also has some duplicate entries. Below, we'll define a function that finds duplicate apps and unique apps in a given dataset. We'll then use that function to find the number of duplicate and unique Android apps in our dataset as well as a few examples of each.

In [9]:
def find_duplicates(dataset):
    duplicate_apps = []
    unique_apps = []
    # loop through dataset, add app to duplicate list if multiples found
    for app in dataset:
        name = app[0]
        if name in unique_apps:
            duplicate_apps.append(name)
        else:
            unique_apps.append(name)
    return [duplicate_apps, unique_apps]

android_duplicates = find_duplicates(android_data)
# print total duplicate and unique apps along with examples for Android dataset
print("Number of duplicate Android apps: ", len(android_duplicates[0]))
print("Examples of duplicate Android apps: ", android_duplicates[0][5:10])
print("Number of unique Android apps: ", len(android_duplicates[1]))
print("Examples of unique Android apps: ", android_duplicates[1][0:5])

Number of duplicate Android apps:  1181
Examples of duplicate Android apps:  ['Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']
Number of unique Android apps:  9659
Examples of unique Android apps:  ['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book']


We can see this duplicate data by creating a function to print duplicate rows, then calling it for one of the duplicate apps we identified above (Slack) to show that there are multiple entries for Slack in our dataset.

In [10]:
def list_duplicates(app_name, dataset):
    # loop through dataset, print all entries for a given app
    for app in dataset:
        name = app[0]
        if name == app_name:
            print(app)

list_duplicates('Slack', android_data)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


By running the `find_duplicates()` function on the dataset for iOS apps, we can see that there are no duplicates in the iOS dataset.

In [11]:
ios_duplicates = find_duplicates(ios_data)
# print total duplicate and unique apps along with examples for iOS dataset
print("Number of duplicate iOS apps: ", len(ios_duplicates[0]))
print("Examples of duplicate iOS apps: ", ios_duplicates[0][0:4])
print("Number of unique iOS apps: ", len(ios_duplicates[1]))
print("Examples of unique iOS apps: ", ios_duplicates[1][0:4])

Number of duplicate iOS apps:  0
Examples of duplicate iOS apps:  []
Number of unique iOS apps:  7197
Examples of unique iOS apps:  ['284882215', '389801252', '529479190', '420009108']


Now we'll remove the duplicates from the Google Play dataset by keeping only the most recent entry for each duplicated app. We'll do this by finding the entry with the highest number of user reviews (found in the `Reviews` column of the dataset at index `[3]`). To continue with our example of Slack, this means that we'll want the Slack entry with 51,510 user reviews, rather than the entry with 51,507 user reviews.

We'll start by creating a dictionary called `reviews_max` that contains the names of all apps in the dataset as keys and their highest number of user reviews as values.

In [12]:
reviews_max = {}
for app in android_data:
    # loop through Android dataset to create dictionary
    name = app[0]
    # convert number of reviews to float
    n_reviews = float(app[3])
    if (name in reviews_max and reviews_max[name] < n_reviews) or name not in reviews_max:
        reviews_max[name] = n_reviews

Now, we'll test whether everything went as expected by finding the length of the `reviews_max` dictionary (which should be 9,659, the same as the number of unique Android apps) and the number of user reviews for Slack (which should be 51,510).

In [13]:
print(len(reviews_max))
print(reviews_max['Slack'])

9659
51510.0


As shown above, everything ran as expected. Now we'll use the dictionary we created to remove  duplicate rows from our Android dataset.

In [14]:
android_clean = []
already_added = []
# loop through dataset to create a dataset free of duplicates
for app in android_data:
    name = app[0]
    # convert number of reviews to float
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)

Below, we'll check the length of the `android_clean` dataset to confirm that our clean dataset has a length of 9,659.

In [15]:
print(len(android_clean))

9659


Having confirmed the correct link, we'll now define a function that confirms whether an app name has more than three characters that aren't common English characters. If so, we'll conclude that that app likely isn't in English (returning `False`).

In [16]:
def is_english(string):
    non_english_chars = 0
    # use Unicode code point of a given character to determine whether it is a common
    # English character. Values above 127 are non-standard characters
    for char in string: 
        if ord(char) > 127: non_english_chars += 1
    # if more than three non-standard characters, app is likely not in English
    if non_english_chars > 3: return False
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


Now, we'll define a function to loop through our datasets to collect all the English-language apps in each dataset and compile them into lists.

In [17]:
def find_english(dataset, name):
    english_apps = []
    for row in dataset:
        if is_english(row[name]):
            english_apps.append(row)
    return english_apps

english_ios = find_english(ios_data, 1)
english_android = find_english(android_data, 0)
# compare numbers of apps in English-language subsets and full datasets
print(len(english_ios), len(english_android))
print(len(ios_data), len(android_data))

6183 10795
7197 10840


Next, we'll follow a similar procedure to identify free apps. As before, we'll loop through each dataset, append free apps to a list, and 

In [18]:
def find_free(dataset, price):
    free_apps = []
    for row in dataset:
        # prices stored as '0.0' in iOS dataset and '0' in Android dataset
        if row[price] == '0.0' or row[price] == '0':
            free_apps.append(row)
    return free_apps

free_ios = find_free(english_ios, 4)
free_android = find_free(english_android, 7)
# compare numbers of apps in free subsets and full datasets
print(len(free_ios), len(free_android))
print(len(ios_data), len(android_data))

3222 9999
7197 10840


Our validation strategy for an app idea has three steps:
1. Build a minimal Android version of the app and add it to the Google Play Store.  
2. If the app receives a good response from users, we develop it further.  
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Ideally, we want to find apps that will be profitable in both markets.

Next, we'll inspect each dataset and identify the columns that could be used to generate frequency tables to determine the most common genres in each market. For iOS, that appears to be the `prime_genre` column (index `12`), while for the Android dataset, the relevant columns appear to be `Category` (index `1`) and `Genres` (index `9`).

In the below cell, we'll define two functions. We can use the first function (`freq_table()`) to create a frequency table of the genres in each dataset, and we can use the second (`display_table()`) to make that frequency table more interpretable. Specifically, the `display_table()` function saves each key–value pair as a tuple, adds each tuple to a list, sorts that list in descending order, and presents the information in a readable format.

In [19]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    table_percentages = {}
    for key in table:
        # calculate percentage and round to two digits
        percentage = round((table[key] / total) * 100, 2)
        table_percentages[key] = percentage 
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

The most common iOS genre by far is games (58.16%), followed by entertainment (7.88%). Other genres are all less than 5%, and most are less than 3%. The majority of apps appear to be designed for entertainment purposes (e.g., games, social networking, music) rather than practical purposes (e.g., education, utilities, productivity). 

Based on the above, it seems that games are the most popular type of app on the App Store. However, that doesn't necessarily mean that they're the most profitable or have the most users. On the contrary, it might be better to develop an app in a less crowded category to minimize competition and increase the chances that our app will show up in users' searches; we'll need more information to determine the most profitable app profile.

In [20]:
display_table(free_ios, 11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


The Android dataset differentiates between category (the broader group of apps to which a given app belongs) and genre (a more niche, specific description of the app's type). The most common Android category is family (17.67%), followed by games (10.59%). Other categories are all less than 8%, and most are less than 4%. The most common genre, however, is tools (7.63%), followed by entertainment (6.0%). Other genres are all roughly 5% or less, and many are less than 1%.

As with the iOS data, we'll need more information to determine the most profitable app profile, but in comparison with the App Store market, the categories and genres in the Android Market seem to be much more evenly distributed. There is no dominant genre, and there's a more even spread of entertainment and practical apps.

In [21]:
display_table(free_android, 1)

FAMILY : 17.67
GAME : 10.59
TOOLS : 7.64
BUSINESS : 4.45
PRODUCTIVITY : 3.95
SPORTS : 3.6
LIFESTYLE : 3.59
COMMUNICATION : 3.59
MEDICAL : 3.54
FINANCE : 3.49
HEALTH_AND_FITNESS : 3.25
PHOTOGRAPHY : 3.12
PERSONALIZATION : 3.08
SOCIAL : 2.92
NEWS_AND_MAGAZINES : 2.77
SHOPPING : 2.57
TRAVEL_AND_LOCAL : 2.46
DATING : 2.27
BOOKS_AND_REFERENCE : 1.99
VIDEO_PLAYERS : 1.7
EDUCATION : 1.51
ENTERTAINMENT : 1.47
MAPS_AND_NAVIGATION : 1.3
FOOD_AND_DRINK : 1.25
HOUSE_AND_HOME : 0.88
LIBRARIES_AND_DEMO : 0.84
AUTO_AND_VEHICLES : 0.82
WEATHER : 0.74
EVENTS : 0.63
ART_AND_DESIGN : 0.61
COMICS : 0.59
PARENTING : 0.58
BEAUTY : 0.53


In [22]:
display_table(free_android, 9)

Tools : 7.63
Entertainment : 6.0
Education : 5.13
Business : 4.45
Productivity : 3.95
Sports : 3.74
Communication : 3.59
Lifestyle : 3.58
Medical : 3.54
Finance : 3.49
Action : 3.41
Health & Fitness : 3.25
Photography : 3.12
Personalization : 3.08
Social : 2.92
News & Magazines : 2.77
Shopping : 2.57
Travel & Local : 2.45
Dating : 2.27
Arcade : 2.0
Books & Reference : 1.99
Simulation : 1.88
Casual : 1.84
Video Players & Editors : 1.68
Maps & Navigation : 1.3
Food & Drink : 1.25
Puzzle : 1.21
Racing : 0.95
Strategy : 0.93
House & Home : 0.88
Role Playing : 0.87
Libraries & Demo : 0.84
Auto & Vehicles : 0.82
Weather : 0.74
Events : 0.63
Adventure : 0.62
Comics : 0.58
Art & Design : 0.54
Beauty : 0.53
Parenting : 0.44
Education;Education : 0.43
Card : 0.41
Educational;Education : 0.38
Casino : 0.38
Trivia : 0.37
Board : 0.35
Educational : 0.33
Word : 0.29
Entertainment;Music & Video : 0.27
Casual;Pretend Play : 0.25
Music : 0.21
Casual;Action & Adventure : 0.2
Racing;Action & Adventure : 

Below, we'll create a frequency table for the free iOS app genres. We'll then calculate the average ratings for each genre.

In [23]:
genres_ios = freq_table(free_ios, 11)
for genre in genres_ios:
    total = 0
    len_genre = 0
    for row in free_ios:
        genre_app = row[11]
        if genre_app == genre:
            user_ratings = float(row[5])
            total += user_ratings
            len_genre += 1
    # round to even number, add comma separators for readability
    avg_ratings = "{:,}".format(round((total / len_genre), 0))
    print(genre, avg_ratings)

Social Networking 71,548.0
Photo & Video 28,442.0
Games 22,789.0
Music 57,327.0
Reference 74,942.0
Health & Fitness 23,298.0
Weather 52,280.0
Utilities 18,684.0
Travel 28,244.0
Shopping 26,920.0
News 21,248.0
Navigation 86,090.0
Lifestyle 16,486.0
Entertainment 14,030.0
Food & Drink 33,334.0
Sports 23,009.0
Book 39,758.0
Finance 31,468.0
Education 7,004.0
Productivity 21,028.0
Business 7,491.0
Catalogs 4,004.0
Medical 612.0


The genres with the greatest number of average user ratings are navigation (*n* = 86,090), reference (*n* = 74,942), and social networking (*n* = 71,548). Games, despite being the most common genre, have only 22,789 average user ratings.

It's likely safe to assume that the high averages for navigation and social networking are due to huge, highly popular apps like Google Maps, Waze, Facebook, and Twitter. Since competing with such dominant apps is a pretty challenging task, it may be a better idea to select a more moderately rated genre like reference, weather, or music. The market for these apps is less crowded, and the apps within them seem to amass a decently high number of user ratings.

Next, we'll follow the same procedure for the Android dataset.

In [24]:
categories_android = freq_table(free_android, 1)
for category in categories_android:
    total = 0
    len_category = 0
    for row in free_android:
        category_app = row[1]
        if category_app == category:
            installs = row[5]
            # strip non-numeric characters and convert to float
            installs = installs.replace('+', '')
            installs = installs.replace(',', '')
            installs = float(installs)
            total += installs
            len_category += 1
    # round to even number, add comma separators for readability
    avg_installs = "{:,}".format(round((total / len_category), 0))
    print(category, avg_installs)

ART_AND_DESIGN 2,038,051.0
AUTO_AND_VEHICLES 647,318.0
BEAUTY 513,152.0
BOOKS_AND_REFERENCE 9,655,197.0
BUSINESS 2,250,454.0
COMICS 950,443.0
COMMUNICATION 90,935,672.0
DATING 1,164,271.0
EDUCATION 5,760,596.0
ENTERTAINMENT 19,516,735.0
EVENTS 253,542.0
FINANCE 2,511,356.0
FOOD_AND_DRINK 2,190,710.0
HEALTH_AND_FITNESS 4,869,226.0
HOUSE_AND_HOME 1,917,187.0
LIBRARIES_AND_DEMO 749,950.0
LIFESTYLE 1,479,957.0
GAME 33,111,303.0
FAMILY 5,784,095.0
MEDICAL 147,563.0
SOCIAL 48,184,459.0
SHOPPING 12,637,504.0
PHOTOGRAPHY 32,321,374.0
SPORTS 4,860,919.0
TRAVEL_AND_LOCAL 27,921,561.0
TOOLS 14,988,277.0
PERSONALIZATION 7,533,233.0
PRODUCTIVITY 35,885,138.0
PARENTING 542,604.0
WEATHER 5,747,142.0
VIDEO_PLAYERS 36,599,010.0
NEWS_AND_MAGAZINES 27,058,831.0
MAPS_AND_NAVIGATION 5,569,698.0


The categories with the greatest number of average user ratings are communication (*n* = 90,935,672), social (*n* = 48,184,459), and video players (*n* = 36,599,010). As above, we can probably assume that the high averages for these apps are for very popular apps like WhatsApp, Messenger, Facebook, and YouTube. The next most popular categories include game (*n* = 33,111,303), photography (*n* = 32,321,374), and entertainment (*n* = 19,516,735). 

Cross-referencing with our results for the iOS store, and keeping in mind our goal of developing an app that could be popular in both markets, we might consider developing an app in the music, books and reference, or photo and video category.